src/webutil/tableState.js

```diff
@@ -0,0 +1,185 @@
+  +length: number,
+  /**
+    A stateful representation of the current page index.
+   */
+  +pageIndex: number,
+  /**
+    A stateful representation of the number of rows per page.
+   */
+  +rowsPerPage: number,
+  /**
+    A stateful representation of the current sort name (the last sort name passed to setSortFn).
+   */
+  +sortName: string,
+  /**
+    A stateful representation of the current order, represented by "asc" or "desc".
+   */
+  +sortOrder: SortOrderTypes,
+  /**
+    A function that changes the currentPage to represent the indicated index.
+   */
+  +setPageIndex: (number) => void,
+  /**
+    A function that sets the rows per page and updates the pageIndex to an equivalent place.
+   */
+  +setRowsPerPage: (number) => void,
+  /**
+    A function that adds a filter to the map of filters or overwrites an existing filter that matches the key.
+    Every filter stored will be applied together as an AND operation (results will only include elements
+    that pass all filters).
+    Sets pageIndex to 0.
+
+    string: A key that represents the filter in the map.
+    function?: A function that returns false if an element should be excluded from the result. Null to remove the filter at the key.
+   */
+  +createOrUpdateFilterFn: (string, null | ((T) => boolean)) => void,
+  /**
+    A function that sorts the data. Current pageIndex is maintained.
+
+    string: A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    function?: A function that returns the property of the element that should be used in the sort comparator.
+   */
+  +setSortFn: (string, PluckFn<T>) => void,
+  /**
+    A function that overwrites the base data and reapplies current sort and filter functions.
+    Sets pageIndex to 0.
+   */
+  +setData: ($ReadOnlyArray<T>) => void,
+|};
+
+/**
+  A builder function for initializing a new TableState for use in a ReactNode.
+  It encapsulates Sort, Pagination, and Filtering.
+ */
+export function useTableState<T>(defaults?: {
+  initialRowsPerPage?: number,
+  initialSort?: {
+    // The starting order of the initial sort, represented by "asc" or "desc".
+    sortOrder: SortOrderTypes,
+    // A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    sortName: string,
+    // A function that returns the property of the element that should be used in the sort comparator.
+    sortFn: PluckFn<T>,
+  },
+}): TableState<T> {
+  defaults = defaults || {};
+  const sortDefaults = defaults.initialSort || {};
+  const [originalArray, setOriginalArray] = useState([]);
+  const [filteredArray, setFilteredArray] = useState([]);
+  const [sortName, setSortName] = useState(sortDefaults.sortName || "");
+  const [sortOrder, setSortOrder] = useState(
+    sortDefaults.sortOrder || SortOrders.ASC
+  );
+  const [pageIndex, setPageIndex] = useState(0);
+  const [rowsPerPage, setRowsPerPageInternal] = useState(
+    defaults.initialRowsPerPage || 20
```

Comment:
Would be good to extract the `20` as a constant `DEFAULT_ROWS_PER_PAGE` before the function definition so its not so much a "magic number".

Reply:
accepted

Reply:
actually, I think I will just make this a required field instead, since there is no real case where you can get away with not providing it.